### PR TITLE
Optionally emit versioned .so

### PR DESCRIPTION
**Pull request by JamesGuthrie**
`pgx` now experimentally supports the option to produce a versioned
shared library. This option is activated by removing the
`module_pathname` parameter in the extension's `.control` file.

When activated, it does the following:

- appends current package version to shared library name
- points generated function SQL to the versioned shared library (instead
  of `MODULE_PATHNAME`)
- replaces `@MODULE_PATHNAME@` placeholder with the path to the versioned
  shared library

Background:

There is an implicit requirement that C extensions maintain ABI
compatibility between versions. Emitting a versioned .so allows for ABI
breaks between versions.

This functionality is experimental and that care must be taken in the
following scenarios:

- when using shared memory
- when using query planner hooks
- when producing SQL schema migrations

### CODE REVIEW

File: cargo-pgx/src/command/install.rs
Comment (eeeebbbbrrrr) [inline cargo-pgx/src/command/install.rs:163]: this incantation also exists over in `pgx_sql.rs`.  How can we centralize them?

File: cargo-pgx/src/command/install.rs
Comment (eeeebbbbrrrr) [inline cargo-pgx/src/command/install.rs:488]: I'm having a hard time visualizing what this is replacing, and where.  Can you clarify?

File: cargo-pgx/src/command/run.rs
Comment (eeeebbbbrrrr) [inline cargo-pgx/src/command/run.rs:51]: Do we need this for `cargo pgx run`?  I'm leaning towards no, but could be convinced otherwise...

Comment (eeeebbbbrrrr) [reply cargo-pgx/src/command/run.rs:51]: If we do do this for `run`, I suppose we'd want to do it for `test` too?

**Review by eeeebbbbrrrr — CHANGES_REQUESTED** [review]
I think overall this is pretty close.

Before merging, I'd like to see the `cargo-pgx/README.md` updated to reflect this new argument and detail what it does any why it might be necessary.

File: cargo-pgx/src/command/run.rs
Comment (JamesGuthrie) [reply cargo-pgx/src/command/run.rs:51]: It's not necessary for `run`, but it sure made testing this change a lot easier for me, as I could just `cargo pgx run` instead of having to `install` and then startup and connect to PG. I can add it for test as well. I don't really see the utility in that, but it would make the CLI API surface a little more consistent. What would you prefer?

File: cargo-pgx/src/command/install.rs
Comment (JamesGuthrie) [reply cargo-pgx/src/command/install.rs:488]: In the `.control` file, one can set the `module_pathname` (path to shared object). With the default pgx extension template, the value of `module_pathname` is set to `$libdir/<extension-name>`. When using the `--versioned-so` feature, this default value for `module_pathname` will be incorrect.

Instead if leaving it in there to break stuff, this change comments it out of the `.control` file during the install phase.

Another option would be to check if a `module_pathname` entry is present in the `.control` file when `--versioned-so` is used, and explicitly throw an error. Actually I think that this approach may be more explicit in what it does.

Comment (JamesGuthrie) [reply cargo-pgx/src/command/install.rs:488]: Another other option would be to make `module_pathname` templated in the control file, for instance setting it to `@MODULE_PATHNAME@`. This would bring it in line with how `default_version` works.

File: cargo-pgx/src/command/install.rs
Comment (JamesGuthrie) [reply cargo-pgx/src/command/install.rs:163]: I don't think that there is any value in trying to do that. There are similarities, yes, but by the time you would have refactored everything to pull out the commonality, the common piece would be this:

```
pub fn build_versioned_so_name(extname: String, extver: String) -> String {
    format!("{}-{}", extname, extver)
}
```

File: cargo-pgx/src/command/run.rs
Comment (JamesGuthrie) [reply cargo-pgx/src/command/run.rs:51]: In the latest iteration I've implemented this for `test`.

**Comment by JamesGuthrie** [comment]
> I think overall this is pretty close.
> 
> Before merging, I'd like to see the `cargo-pgx/README.md` updated to reflect this new argument and detail what it does any why it might be necessary.

I've updated the docs. I'm having difficulty seeing where somebody who is unfamiliar with these concepts will have difficulty, so if there's anything that you find unclear please let me know and I'll take another pass.

File: cargo-pgx/src/command/install.rs
Comment (JamesGuthrie) [reply cargo-pgx/src/command/install.rs:163]: I see. I've fixed this in the most recent iteration.

File: cargo-pgx/src/command/install.rs
Comment (eeeebbbbrrrr) [reply cargo-pgx/src/command/install.rs:489]: Based on Postgres' docs for `module_pathname`, it really doesn't mean anything unless `MODULE_PATHNAME` exists in any of the extension sql scripts.  And with this PR, I guess we wouldn't.

Here's an alternative thought...

What if instead of having the `--versioned-so` CLI arg, we instead look for `module_pathname` in the .control file.  If it's there, we do the default behavior, if it's not, then we do the versioned-so behavior?  

I'm still not 100% convinced that if you want versioned .so's that you'd want to turn them off from time to time or from cargo pgx command to command.

Thoughts?

Comment (JamesGuthrie) [reply cargo-pgx/src/command/install.rs:489]: > And with this PR, I guess we wouldn't.

Yes and no. By user error, it's possible that `MODULE_PATHNAME` would still be set in an SQL override.

> What if instead of having the --versioned-so CLI arg, we instead look for module_pathname in the .control file. If it's there, we do the default behavior, if it's not, then we do the versioned-so behavior?

That's an elegant way to "hide" the optionality. I kinda like it. I can't think of a scenario which it causes problems with, but maybe that's my lack of creativity. One downside it that it's quite implicit.

> I'm still not 100% convinced that if you want a versioned .so that there'd also be times when you wouldn't. That doesn't jive in my little brain.

I think that you're right on this.

Comment (JamesGuthrie) [reply cargo-pgx/src/command/install.rs]: > You wanna go ahead and make these changes? Sorry for the churn, but we really need to get this right so it's not surprising to users.

Yup, I've changed it to work as you described. Please take another look.

File: pgx-examples/versioned_so/versioned_so.control
Comment (eeeebbbbrrrr) [inline pgx-examples/versioned_so/versioned_so.control:4]: Just for clarity, add a commented out `module_pathname` property here, with some explanation as to what it means to **not** have defined.

**Review by eeeebbbbrrrr — CHANGES_REQUESTED** [review]

File: pgx-examples/versioned_so/versioned_so.control
Comment (eeeebbbbrrrr) [reply pgx-examples/versioned_so/versioned_so.control:4]: Also, please add a run line for the example in the appropriate place around here:  https://github.com/tcdi/pgx/blob/c4ce4b9a5d74d589345fe95f3175a9e656dc6b3f/.github/workflows/tests.yml#L135

Comment (eeeebbbbrrrr) [reply pgx-examples/versioned_so/versioned_so.control:4]: Also, also, please add the example to the top-level workspace here:  https://github.com/tcdi/pgx/blob/master/Cargo.toml

**Comment by eeeebbbbrrrr** [comment]
@JamesGuthrie this is super nice now.  Great work and thanks for sticking through the process.  When you address those three minor comments above this will be ready to merge.  

Oh, please retarget the PR to the `develop` branch as well.  We're weird and don't merge directly into master, unless you're me and are prepared to be chastised by @Hoverbear for doing it.  ;)

File: pgx-examples/versioned_so/versioned_so.control
Comment (JamesGuthrie) [reply pgx-examples/versioned_so/versioned_so.control]: Done. And added the example to the GitHub workflow.

**Comment by JamesGuthrie** [comment]
> @JamesGuthrie this is super nice now. Great work and thanks for sticking through the process. When you address those three minor comments above this will be ready to merge.
> 
> Oh, please retarget the PR to the `develop` branch as well. We're weird and don't merge directly into master, unless you're me and are prepared to be chastised by @Hoverbear for doing it. ;)

No problem, thanks for the quick review! I've fixed the three things above and retargeted against develop.

**Review by eeeebbbbrrrr — APPROVED** [review]

**Comment by eeeebbbbrrrr** [comment]
Approved.  Lets see what CI says and then I'll merge it.

And as promised, I'm happy to put out a v0.4.4 this week so y'all can get back to your real work.

File: cargo-pgx/README.md
Comment (Hoverbear) [inline cargo-pgx/README.md:732]: So the 0.0.1 schema would contain: 

```sql

```SQL
CREATE OR REPLACE FUNCTION "hello_extension"() RETURNS text /* &str */
STRICT
LANGUAGE c /* Rust */
AS '$libdir/extension-0.0.1', 'hello_extension_wrapper';
```

Still?

**Comment by eeeebbbbrrrr** [comment]
hmm, the new `versioned_so` test failed:  https://github.com/tcdi/pgx/runs/6161094219?check_suite_focus=true

Error message sucks.  Did it work for you locally?

File: cargo-pgx/README.md
Comment (Hoverbear) [inline cargo-pgx/README.md:729]: Does versioned shared-object support also make the output `.so` have a different name?

Comment (eeeebbbbrrrr) [reply cargo-pgx/README.md:729]: > Does versioned shared-object support also make the output .so have a different name?

Yup

File: pgx-examples/versioned_so/versioned_so.control
Comment (Hoverbear) [inline pgx-examples/versioned_so/versioned_so.control:3]: I love how you left this comment to explain it for readers. :)

**Review by Hoverbear — COMMENTED** [review]
Some minor readme expansions requested, but otherwise I'm happy with how this turned out!

File: cargo-pgx/README.md
Comment (JamesGuthrie) [reply cargo-pgx/README.md:729]: I've mentioned it here in the readme.

File: cargo-pgx/README.md
Comment (JamesGuthrie) [reply cargo-pgx/README.md:732]: I've added and explicit example.

**Comment by JamesGuthrie** [comment]
Okay, managed to get the build to work locally. Let's hope that CI accepts it.

**Review by Hoverbear — APPROVED** [review]

**Comment by Hoverbear** [comment]
This is great James. :)

**Comment by eeeebbbbrrrr** [comment]
Thanks a bunch, @JamesGuthrie!  Merged.  We'll get it out in a new release in the next few days.

File: pgx-examples/versioned_so/.cargo/config
Comment (Hoverbear) [inline pgx-examples/versioned_so/.cargo/config:3]: We didn't catch this during review, but this should have been removed as it causes some conflicts with the workspace override. @eeeebbbbrrrr is fixing this up before release.

Comment (eeeebbbbrrrr) [reply pgx-examples/versioned_so/.cargo/config:3]: yah.  totally our bad.

In a workspace crate environment like pgx is, you **only** need the `.cargo/config` at the root workspace crate.  If it's in any "child" crates, the compiler goes stupid and you spend all day trying to figure it out and then you're thankful @Hoverbear is on your team!

**Pull request merged**
